sql/exec: TestHashRouterRandom failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/84e32553521695cf97424f2cc72019643de5cf91

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestHashRouterRandom PKG=github.com/cockroachdb/cockroach/pkg/sql/exec TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1257906&tab=buildLog

```

panic: query execution canceled

goroutine 2108 [running]:
github.com/cockroachdb/cockroach/pkg/sql/exec.(*CancelChecker).checkEveryCall(0xc0038301d0, 0x17b31a0, 0xc003543800)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/cancel_checker.go:71 +0x6e
github.com/cockroachdb/cockroach/pkg/sql/exec.(*CancelChecker).check(0xc0038301d0, 0x17b31a0, 0xc003543800)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/cancel_checker.go:57 +0x57
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashTable).rehash(0xc003830028, 0x17b31a0, 0xc003543800, 0xc003832000, 0x400, 0x400, 0x0, 0x6, 0x17c6e80, 0xc00355b410, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/hashjoiner.eg.go:209 +0x1673
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashRouter).processNextBatch(0xc003830000, 0x17b31a0, 0xc003543800, 0x1015e9718)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:374 +0x186
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashRouter).run(0xc003830000, 0x17b31a0, 0xc003543800)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:351 +0x105
github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1.2(0xc003830000, 0x17b31a0, 0xc003543800, 0xc000290150)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:731 +0x3f
created by github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:730 +0x4fc

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0001a8000, 0x1554620, 0x14, 0x15e6e08, 0x494401)
	/usr/local/go/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc0001a8800)
	/usr/local/go/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc0001a8800, 0xc00049fde8)
	/usr/local/go/src/testing/testing.go:827 +0xbf
testing.runTests(0xc000224ac0, 0x24602a0, 0x35, 0x35, 0x1c4ea10)
	/usr/local/go/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc00002b900, 0x0)
	/usr/local/go/src/testing/testing.go:1034 +0x165
github.com/cockroachdb/cockroach/pkg/sql/exec.TestMain(0xc00002b900)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/main_test.go:26 +0x30
main.main()
	_testmain.go:176 +0x13d

goroutine 20 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 21 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1226 +0xf2
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:629 +0x11e

goroutine 22 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:636 +0xab
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:630 +0x136

goroutine 1651 [chan receive]:
testing.(*T).Run(0xc0001a8200, 0xc0004da070, 0x62, 0xc0001ac750, 0x6)
	/usr/local/go/src/testing/testing.go:879 +0x383
github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom(0xc0001a8000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:691 +0x7a9
testing.tRunner(0xc0001a8000, 0x15e6e08)
	/usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x35c

goroutine 2087 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc003542dd0, 0xc000000000)
	/usr/local/go/src/runtime/sema.go:510 +0xeb
sync.(*Cond).Wait(0xc003542dc0)
	/usr/local/go/src/sync/cond.go:56 +0x92
github.com/cockroachdb/cockroach/pkg/sql/exec.(*routerOutputOp).Next(0xc00382ea80, 0x17b31e0, 0xc0000c4010, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:107 +0x92
github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1.1(0xc00382c2c0, 0x2a, 0x2a, 0x17b31e0, 0xc0000c4010, 0xc000290128, 0xc00384c000, 0x2a, 0x2a, 0xc000290150, ...)
```

SHA: https://github.com/cockroachdb/cockroach/commits/84e32553521695cf97424f2cc72019643de5cf91

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestHashRouterRandom PKG=github.com/cockroachdb/cockroach/pkg/sql/exec TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1257907&tab=buildLog

```

panic: query execution canceled

goroutine 3843 [running]:
github.com/cockroachdb/cockroach/pkg/sql/exec.(*CancelChecker).checkEveryCall(0xc00018a450, 0x21ef0e0, 0xc0002380c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/cancel_checker.go:71 +0x95
github.com/cockroachdb/cockroach/pkg/sql/exec.(*CancelChecker).check(0xc00018a450, 0x21ef0e0, 0xc0002380c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/cancel_checker.go:57 +0xba
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashTable).rehash(0xc00018a2a8, 0x21ef0e0, 0xc0002380c0, 0xc000472000, 0x400, 0x400, 0x0, 0x6, 0x220afe0, 0xc0117a2000, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/hashjoiner.eg.go:209 +0x235e
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashRouter).processNextBatch(0xc00018a280, 0x21ef0e0, 0xc0002380c0, 0xc000489340)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:374 +0x354
github.com/cockroachdb/cockroach/pkg/sql/exec.(*hashRouter).run(0xc00018a280, 0x21ef0e0, 0xc0002380c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:351 +0x1ce
github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1.2(0xc00018a280, 0x21ef0e0, 0xc0002380c0, 0xc0002b8090)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:731 +0x4d
created by github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:730 +0x6b6

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00051c000, 0x1d50759, 0x14, 0x1de30e0, 0xc000589b01)
	/usr/local/go/src/testing/testing.go:879 +0x692
testing.runTests.func1(0xc00051c000)
	/usr/local/go/src/testing/testing.go:1119 +0xa9
testing.tRunner(0xc00051c000, 0xc000589d68)
	/usr/local/go/src/testing/testing.go:827 +0x163
testing.runTests(0xc000224000, 0x3077360, 0x35, 0x35, 0x0)
	/usr/local/go/src/testing/testing.go:1117 +0x4ef
testing.(*M).Run(0xc00058a000, 0x0)
	/usr/local/go/src/testing/testing.go:1034 +0x2ef
github.com/cockroachdb/cockroach/pkg/sql/exec.TestMain(0xc00058a000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/main_test.go:26 +0x3e
main.main()
	_testmain.go:176 +0x222

goroutine 20 [syscall]:
os/signal.signal_recv(0x477831)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x30
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x4f

goroutine 21 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1226 +0x13c
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:629 +0x1e4

goroutine 22 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:636 +0xc7
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:630 +0x1fc

goroutine 4046 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0005f0cd0, 0xc000000000)
	/usr/local/go/src/runtime/sema.go:510 +0xeb
sync.(*Cond).Wait(0xc0005f0cc0)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/sql/exec.(*routerOutputOp).Next(0xc0002ed200, 0x21ef120, 0xc0000c4010, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers.go:107 +0xd3
github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1.1(0xc0001b5000, 0xef, 0xef, 0x21ef120, 0xc0000c4010, 0xc0002b8080, 0xc011797800, 0xef, 0xef, 0xc0002b8090, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:717 +0xfd
created by github.com/cockroachdb/cockroach/pkg/sql/exec.TestHashRouterRandom.func1.1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/routers_test.go:715 +0x5e1

goroutine 3949 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0005ed050, 0xc000000000)
	/usr/local/go/src/runtime/sema.go:510 +0xeb
sync.(*Cond).Wait(0xc0005ed040)
	/usr/local/go/src/sync/cond.go:56 +0x8e
```

--- a/pkg/sql/exec/routers.go
+++ b/pkg/sql/exec/routers.go
@@ -348,7 +348,16 @@ func (r *hashRouter) run(ctx context.Context) {
 			}
 		}
 
-		if done := r.processNextBatch(ctx); done {
+		var done bool
+		if err := CatchVectorizedRuntimeError(func() {
+			done = r.processNextBatch(ctx)
+		}); err != nil {
+			// TODO(asubiotto): Propagate this error through metadata. Think about
+			// semantics.
+			cancelOutputs()
+			return
+		}
+		if done {
 			// The input was done and we have notified the routerOutputs that there
 			// is no more data.
 			return

